File: julia/resistor-color-trio/resistor-color-trio.jl
module Resistor
  Value = Dict(
    "black" => 0, "brown" => 1, "red" => 2, "orange" => 3, "yellow" => 4,
    "green" => 5, "blue" => 6, "violet" => 7, "grey" => 8, "white" => 9
  )
  Unit = "ohms"
  Prefixes = ["", "kilo", "mega", "giga"]

  function value(colours)
    @assert length(colours) ≥ 3 "must pass at least 3 colours"
    @assert all(c -> haskey(Value, c), colours) "not all colours are valid"

    first, second, third = [Value[c] for c ∈ colours]
    (10 * first + second) * 10 ^ third
  end

  function to_string(value)
    idx = 1
    while value > 0 && mod(value, 1000) == 0
      value ÷= 1000
      idx += 1
    end
    string(value, " ", Prefixes[idx], Unit)
  end
end


function label(colors)
  Resistor.to_string(Resistor.value(colors))
end

File: nim/resistor-color-trio/resistor_color_trio.nim
import math

type
  ResistorColor* = enum
    Black, Brown, Red, Orange, Yellow,
    Green, Blue, Violet, Grey, White

proc colorCode(color: ResistorColor): int = color.ord

proc value(colors: openArray[ResistorColor]): int =
  assert colors.len >= 2, "not enough colors"
  10 * colors[0].colorCode + colors[1].colorCode

proc label*(colors: openArray[ResistorColor]): (int, string) =
  assert colors.len == 3, "incorrect input"
  var resistance = value([colors[0], colors[1]])
  resistance *= 10^colors[2].colorCode

  var magnitude = 0
  while resistance mod 1000 == 0:
    resistance = resistance div 1000
    magnitude.inc
  var prefix = ["", "kilo", "mega", "giga"][magnitude]

  (resistance, prefix&"ohms")
